# Проект "предсказание стоимости жилья (Калифорния, 1990)"

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc">
<ul class="toc-item">
<li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span><u>Описание проекта</u></a></span>
<ul class="toc-item">
<li><span><a href="#Задача-и-контекст" data-toc-modified-id="Задача-и-контекст-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Задача и контекст</a></span></li>
<li><span><a href="#Ключевые-цели" data-toc-modified-id="Ключевые-цели-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Ключевые цели</a></span></li>
<li><span><a href="#Практическое-применение" data-toc-modified-id="Практическое-применение-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Практическое применение</a></span></li>
<li><span><a href="#Бизнес-ценность" data-toc-modified-id="Бизнес-ценность-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Бизнес-ценность</a></span></li>
</ul>
</li>
<li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span><u>Описание данных</u></a></span>
<ul class="toc-item">
<li><span><a href="#Общая-информация-о-датасете" data-toc-modified-id="Общая-информация-о-датасете-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Общая информация о датасете</a></span></li>
<li><span><a href="#Структура-данных" data-toc-modified-id="Структура-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Структура данных</a></span></li>
<li><span><a href="#Категориальные-признаки" data-toc-modified-id="Категориальные-признаки-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Категориальные признаки</a></span></li>
<li><span><a href="#Целевая-переменная" data-toc-modified-id="Целевая-переменная-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Целевая переменная</a></span></li>
<li><span><a href="#Метрики-оценки-качества" data-toc-modified-id="Метрики-оценки-качества-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Метрики оценки качества</a></span></li>
<li><span><a href="#Ожидаемые-сложности" data-toc-modified-id="Ожидаемые-сложности-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Ожидаемые сложности</a></span></li>
</ul>
</li>
<li><span><a href="#Загрузка-библиотек" data-toc-modified-id="Загрузка-библиотек-3"><span class="toc-item-num">3&nbsp;&nbsp;</span><u>Загрузка библиотек</u></a></span></li>
<li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span><u>Подготовка данных</u></a></span>
<ul class="toc-item">
<li><span><a href="#Инициализация-локальной-Spark-сессии" data-toc-modified-id="Инициализация-локальной-Spark-сессии-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Инициализация локальной Spark-сессии</a></span></li>
<li><span><a href="#Вывод-типов-данных" data-toc-modified-id="Вывод-типов-данных-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Вывод типов данных</a></span></li>
</ul>
</li>
<li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-5"><span class="toc-item-num">5&nbsp;&nbsp;</span><u>Предобработка данных</u></a></span>
<ul class="toc-item">
<li><span><a href="#Замена-пропусков" data-toc-modified-id="Замена-пропусков-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Замена пропусков</a></span></li>
<li><span><a href="#Преобразование-категориальных-колонок" data-toc-modified-id="Преобразование-категориальных-колонок-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Преобразование категориальных колонок</a></span></li>
<li><span><a href="#Трансформация-числовых-признаков" data-toc-modified-id="Трансформация-числовых-признаков-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Трансформация числовых признаков</a></span></li>
<li><span><a href="#Объединение-признаков" data-toc-modified-id="Объединение-признаков-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Объединение признаков</a></span></li>
</ul>
</li>
<li><span><a href="#Обучение-моделей-и-предсказания" data-toc-modified-id="Обучение-моделей-и-предсказания-6"><span class="toc-item-num">6&nbsp;&nbsp;</span><u>Обучение моделей и предсказания</u></a></span>
<ul class="toc-item">
<li><span><a href="#Модель-1-все-данные" data-toc-modified-id="Модель-1-все-данные-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Модель 1 (все данные)</a></span></li>
<li><span><a href="#Модель-2-только-числовые-переменные" data-toc-modified-id="Модель-2-только-числовые-переменные-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Модель 2 (только числовые переменные)</a></span></li>
<li><span><a href="#Модель-3-Random-Forest" data-toc-modified-id="Модель-3-Random-Forest-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Модель 3 (Random Forest)</a></span></li>
<li><span><a href="#Анализ-результатов" data-toc-modified-id="Анализ-результатов-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Анализ результатов</a></span></li>
<li><span><a href="#Сравнение-моделей" data-toc-modified-id="Сравнение-моделей-6.5"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>Сравнение моделей</a></span></li>
</ul>
</li>
<li><span><a href="#Общий-вывод-проекта" data-toc-modified-id="Общий-вывод-проекта-7"><span class="toc-item-num">7&nbsp;&nbsp;</span><u>Общий вывод проекта</u></a></span>
<ul class="toc-item">
<li><span><a href="#Ключевые-результаты" data-toc-modified-id="Ключевые-результаты-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Ключевые результаты</a></span></li>
<li><span><a href="#Сравнение-моделей-1" data-toc-modified-id="Сравнение-моделей-1-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>Сравнение моделей</a></span></li>
<li><span><a href="#Практические-рекомендации" data-toc-modified-id="Практические-рекомендации-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>Практические рекомендации</a></span></li>
<li><span><a href="#Бизнес-ценность-1" data-toc-modified-id="Бизнес-ценность-1-7.4"><span class="toc-item-num">7.4&nbsp;&nbsp;</span>Бизнес-ценность</a></span></li>
<li><span><a href="#Ограничения-и-возможности-улучшения" data-toc-modified-id="Ограничения-и-возможности-улучшения-7.5"><span class="toc-item-num">7.5&nbsp;&nbsp;</span>Ограничения и возможности улучшения</a></span></li>
<li><span><a href="#Заключение" data-toc-modified-id="Заключение-7.6"><span class="toc-item-num">7.6&nbsp;&nbsp;</span>Заключение</a></span></li>
</ul>
</li>
</ul>
</div>

## __1. Описание проекта__

**Задача:** Построение модели линейной регрессии для прогнозирования стоимости жилья в Калифорнии на основе данных 1990 года.

**Контекст:** Проект представляет собой классическую задачу регрессии, где необходимо предсказать медианную стоимость домов в жилых массивах на основе географических, демографических и экономических характеристик.

**Ключевые цели:**
1. Разработка точной модели для оценки стоимости недвижимости
2. Анализ факторов, наиболее сильно влияющих на цены на жильё
3. Оценка качества модели с помощью стандартных метрик регрессии

**Практическое применение:**
- Для **риелторов и агентств недвижимости** — автоматизация оценки объектов
- Для **инвесторов** — выявление перспективных районов для вложений
- Для **банков** — оценка залоговой стоимости при ипотечном кредитовании
- Для **муниципалитетов** — анализ рынка жилья и планирование развития районов

**Бизнес-ценность:**
- Снижение субъективности в оценке недвижимости
- Ускорение процесса оценки объектов
- Повышение точности прогнозов стоимости
- Возможность массовой оценки больших объёмов недвижимости


## __2. Описание данных__

**Название датасета:** California Housing Dataset (1990)

**Источник:** Перепись населения США 1990 года

**Объём данных:** ~20,000 записей (жилых массивов)

**Структура данных:**

| Признак | Тип данных | Описание | Единицы измерения |
|:---------|:------------|:----------|:-------------------|
| `longitude` | float64 | Долгота географического центра | градусы |
| `latitude` | float64 | Широта географического центра | градусы |
| `housing_median_age` | float64 | Медианный возраст зданий | годы |
| `total_rooms` | float64 | Общее количество комнат | штуки |
| `total_bedrooms` | float64 | Общее количество спален | штуки |
| `population` | float64 | Количество жителей | человек |
| `households` | float64 | Количество домовладений | единицы |
| `median_income` | float64 | Медианный доход жителей | десятки тыс. $ |
| `ocean_proximity` | object | Близость к океану | категория |
| `median_house_value` | float64 | Медианная стоимость дома | $ (Целевой признак) |

**Категории признака `ocean_proximity`:**
- `INLAND` — в глубине материка
- `NEAR OCEAN` — рядом с океаном
- `NEAR BAY` — рядом с заливом
- `ISLAND` — остров
- `<1H OCEAN` — менее 1 часа до океана

**Целевая переменная:**
- `median_house_value` — медианная стоимость дома в жилом массиве
- Диапазон значений: от $15,000 до $500,000
- Необходимо предсказать на основе остальных признаков

**Метрики оценки качества:**
1. **RMSE (Root Mean Square Error)** — среднеквадратичная ошибка
   - Измеряется в тех же единицах, что и целевая переменная ($)
   - Чувствительна к выбросам
   
2. **MAE (Mean Absolute Error)** — средняя абсолютная ошибка
   - Более устойчива к выбросам
   - Измеряется в долларах
   
3. **R² (Coefficient of Determination)** — коэффициент детерминации
   - Показывает долю объяснённой дисперсии
   - Диапазон: от 0 до 1 (чем ближе к 1, тем лучше)

**Ожидаемые сложности:**
1. Наличие выбросов в данных
2. Нелинейные зависимости между признаками
3. Влияние географического расположения на стоимость
4. Необходимость обработки категориальных признаков
5. Разные масштабы числовых признаков

# Предсказание стоимости жилья

В проекте вам нужно обучить модель линейной регрессии на данных о жилье в Калифорнии в 1990 году. На основе данных нужно предсказать медианную стоимость дома в жилом массиве. Обучите модель и сделайте предсказания на тестовой выборке. Для оценки качества модели используйте метрики RMSE, MAE и R2.

## 3. __Загрузка библиотек__

In [1]:
# Базовые библиотеки для работы с данными
import pandas as pd 
import numpy as np

# Библиотеки для работы со Spark
import pyspark
from pyspark.sql import SparkSession
from pyspark.sql.types import *
import pyspark.sql.functions as F
from pyspark.sql.functions import col
from pyspark.sql.functions import mean


# Модели машинного обучения
from pyspark.ml.feature import StringIndexer, VectorAssembler, StandardScaler
from pyspark.ml.regression import LinearRegression
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.regression import RandomForestRegressor

print("Библиотеки успешно загружены")

Библиотеки успешно загружены


## __4. Подготовка данных__

### 4.1 Инициализация локальной Spark-сессии

In [2]:
pyspark_version = pyspark.__version__
if int(pyspark_version[:1]) == 3:
    from pyspark.ml.feature import OneHotEncoder    
elif int(pyspark_version[:1]) == 2:
    from pyspark.ml.feature import OneHotEncodeEstimator
        
RANDOM_SEED = 2022

spark = SparkSession.builder \
                    .master("local") \
                    .appName("Housing - Linear regression") \
                    .getOrCreate()

df = spark.read.option('header', 'true').csv('/datasets/housing.csv', inferSchema = True)

### 4.2 Вывод типов данных

In [3]:
df.printSchema()

root
 |-- longitude: double (nullable = true)
 |-- latitude: double (nullable = true)
 |-- housing_median_age: double (nullable = true)
 |-- total_rooms: double (nullable = true)
 |-- total_bedrooms: double (nullable = true)
 |-- population: double (nullable = true)
 |-- households: double (nullable = true)
 |-- median_income: double (nullable = true)
 |-- median_house_value: double (nullable = true)
 |-- ocean_proximity: string (nullable = true)



________________________________
__<u>Промежуточный вывод:<u>__

**1. Инициализация Spark-сессии:**
- Успешно создана локальная Spark-сессия с именем "Housing - Linear regression"
- Определена совместимость с версиями Spark 2 и 3 через условный импорт OneHotEncoder
- Установлен случайный seed для воспроизводимости результатов: `RANDOM_SEED = 2022`

**2. Загрузка данных:**
- Данные успешно загружены из файла `/datasets/housing.csv`
- Использованы параметры:
  - `header='true'` — первая строка содержит названия столбцов
  - `inferSchema=True` — автоматическое определение типов данных
- Загружено 10 признаков датасета о жилье в Калифорнии

**3. Структура данных:**
- **Всего признаков**: 10
- **Числовые признаки (double)**: 9
  - Географические: `longitude`, `latitude`
  - Демографические: `housing_median_age`, `population`, `households`
  - Характеристики жилья: `total_rooms`, `total_bedrooms`
  - Экономические: `median_income`, `median_house_value` (целевая переменная)
- **Категориальный признак (string)**: 1
  - `ocean_proximity` — требует кодирования перед использованием в модели

**4. Статус данных:**
- Все поля допускают NULL значения (`nullable = true`)
- Типы данных определены корректно для дальнейшего анализа


## __5. Предобработка данных__

### 5.1 Замена пропусков

In [4]:
df.describe().toPandas()

summary            longitude           latitude  housing_median_age  \
0   count                20640              20640               20640   
1    mean  -119.56970445736148   35.6318614341087  28.639486434108527   
2  stddev    2.003531723502584  2.135952397457101   12.58555761211163   
3     min              -124.35              32.54                 1.0   
4     max              -114.31              41.95                52.0   

          total_rooms      total_bedrooms          population  \
0               20640               20433               20640   
1  2635.7630813953488   537.8705525375618  1425.4767441860465   
2  2181.6152515827944  421.38507007403115    1132.46212176534   
3                 2.0                 1.0                 3.0   
4             39320.0              6445.0             35682.0   

          households       median_income  median_house_value ocean_proximity  
0              20640               20640               20640           20640  
1  499.5396802325581  3.8706710029070246  206855.81690891474            None  
2  382.3297528316098   1.899821717945263  115395.61587441359            None  
3                1.0              0.4999             14999.0       <1H OCEAN  
4             6082.0             15.0001            500001.0      NEAR OCEAN

In [5]:
display("Пропуски до обработки:")
for column in df.columns:
    null_count = df.filter(col(column).isNull()).count()
    print(f"{column}: {null_count} пропусков")

'Пропуски до обработки:'

longitude: 0 пропусков
latitude: 0 пропусков
housing_median_age: 0 пропусков
total_rooms: 0 пропусков
total_bedrooms: 207 пропусков
population: 0 пропусков
households: 0 пропусков
median_income: 0 пропусков
median_house_value: 0 пропусков
ocean_proximity: 0 пропусков


__Комментарий:__ Пропуск в колонке `total_bedrooms` допустимо заменить средним

In [6]:
mean_value=df.select(mean('total_bedrooms')).collect()[0][0]
df=df.fillna(mean_value, subset='total_bedrooms')

display("Пропуски после обработки:")
for column in df.columns:
    null_count = df.filter(col(column).isNull()).count()
    print(f"{column}: {null_count} пропусков")

'Пропуски после обработки:'

longitude: 0 пропусков
latitude: 0 пропусков
housing_median_age: 0 пропусков
total_rooms: 0 пропусков
total_bedrooms: 0 пропусков
population: 0 пропусков
households: 0 пропусков
median_income: 0 пропусков
median_house_value: 0 пропусков
ocean_proximity: 0 пропусков


### 5.2 Преобразование категориальных колонок

__Комментарий:__ Распределим колонки по типам данных

In [7]:
num_cols=['longitude','latitude','housing_median_age', 'total_rooms', 'total_bedrooms', 'total_rooms', 
     'households', 'median_income']
cat_cols=['ocean_proximity']
target='median_house_value'

In [8]:
all_features=[num_cols, cat_cols]

__Комментарий:__ Поделим выборку на обучающую и тестовую, чтобы стандартизировать и масштабировать их отдельно

In [9]:
train_df, test_df = df.randomSplit([0.8, 0.2], seed=RANDOM_SEED)

In [10]:
indexer= StringIndexer(inputCols=cat_cols,
                      outputCols=[c+'_idx' for c in cat_cols],
                      handleInvalid="keep")

indexer_model=indexer.fit(train_df)

train_df=indexer_model.transform(train_df)

cols = [c for c in train_df.columns for i in cat_cols if (c.startswith(i))]

train_df.select(cols).show(3) 

+---------------+-------------------+
|ocean_proximity|ocean_proximity_idx|
+---------------+-------------------+
|     NEAR OCEAN|                2.0|
|     NEAR OCEAN|                2.0|
|     NEAR OCEAN|                2.0|
+---------------+-------------------+
only showing top 3 rows



In [11]:
test_df=indexer_model.transform(test_df)

cols=[c for c in test_df.columns for i in cat_cols if (c.startswith(i))]

test_df.select(cols).show(3)

+---------------+-------------------+
|ocean_proximity|ocean_proximity_idx|
+---------------+-------------------+
|     NEAR OCEAN|                2.0|
|     NEAR OCEAN|                2.0|
|     NEAR OCEAN|                2.0|
+---------------+-------------------+
only showing top 3 rows



In [12]:
encoder = OneHotEncoder(
    inputCols=[c + '_idx' for c in cat_cols],
    outputCols=[c + '_ohe' for c in cat_cols]
)

encoder_model=encoder.fit(train_df)
train_df=encoder_model.transform(train_df)
test_df=encoder_model.transform(test_df)

### 5.3 Трансформация числовых признаков

In [13]:
assembler_num = VectorAssembler(
    inputCols=num_cols,
    outputCol="numerical_features"
)

train_df=assembler_num.transform(train_df)
test_df=assembler_num.transform(test_df)

In [14]:
scaler = StandardScaler(
    inputCol="numerical_features",
    outputCol="scaled_numerical_features",
    withMean=True
)

scaler_model=scaler.fit(train_df)
train_df=scaler_model.transform(train_df)
test_df=scaler_model.transform(test_df)

### 5.4 Объединение признаков

In [15]:
assembler_all = VectorAssembler(
    inputCols=['scaled_numerical_features', 'ocean_proximity_ohe'],  # ← ВАЖНО: используем МАСШТАБИРОВАННЫЕ признаки!
    outputCol="all_features"  
)

train_df = assembler_all.transform(train_df)
test_df = assembler_all.transform(test_df)


_______________________________________________________________________

<h1 style="font-size: 20px; color: #000000; font-weight: bold;">
    Вывод по разделу
</h1>

**Обработка пропусков**

- Найдена одна колонка с пропусками: `total_bedrooms` (207 пропусков из 20640 записей, ~1%)
- Стратегия заполнения: заменены на среднее значение (537.87)
- Результат: все пропуски устранены, данные готовы для анализа

**Преобразование категориальных признаков**

- Категориальный признак: `ocean_proximity` (5 уникальных значений)
- Кодирование выполнено в 2 этапа:
  1. `StringIndexer`: преобразовал текстовые значения в числовые индексы
  2. `OneHotEncoder`: создал бинарные признаки для каждого значения
    
- *Важная настройка*: `handleInvalid="keep"` предотвращает ошибки при новых категориях

**Обработка числовых признаков**

- Нормализация выполнена для 8 числовых признаков
- Использован StandardScaler: центрирование (withMean=True) и масштабирование

**Объединение признаков**

- Созданы два вектора признаков:
  1. `scaled_numerical_features` - нормализованные числовые признаки
  2. `ocean_proximity_ohe` - закодированные категориальные признаки
- Финальный вектор: `all_features` объединяет оба типа признаков


## __6. Обучение моделей и предсказания__

### 6.1 Модель 1(все данные)

__Коментарий:__ Обучим модель линейной регрессии на всех данных и сделаем предсказание

In [16]:
lr_1=LinearRegression(labelCol=target, featuresCol="all_features")
model_1=lr_1.fit(train_df)

pred_1=model_1.transform(test_df)
predicted_labels_1=pred_1.select('median_house_value', 'prediction').withColumnRenamed("prediction", "pred_1")

predicted_labels_1.show()

26/01/30 14:13:06 WARN Instrumentation: [0d0eaf77] regParam is zero, which might cause numerical instability and overfitting.
26/01/30 14:13:06 WARN BLAS: Failed to load implementation from: com.github.fommil.netlib.NativeSystemBLAS
26/01/30 14:13:06 WARN BLAS: Failed to load implementation from: com.github.fommil.netlib.NativeRefBLAS
26/01/30 14:13:07 WARN LAPACK: Failed to load implementation from: com.github.fommil.netlib.NativeSystemLAPACK
26/01/30 14:13:07 WARN LAPACK: Failed to load implementation from: com.github.fommil.netlib.NativeRefLAPACK


+------------------+------------------+
|median_house_value|            pred_1|
+------------------+------------------+
|          103600.0|160126.65542057715|
|           50800.0| 223368.4219886344|
|           58100.0|143829.92684542201|
|           68400.0|146816.52970568766|
|           72200.0|167967.34580237814|
|           67000.0|162559.42492093123|
|           81300.0|155617.20670023467|
|           70500.0|162410.01754545048|
|           60000.0| 140445.0082957833|
|          109400.0|180011.49533935357|
|           74100.0| 149319.3281802002|
|           74700.0| 159815.8991947663|
|           90000.0|206951.80061553698|
|          104200.0|201706.76466416032|
|           74100.0| 155816.0509583049|
|           67500.0|151018.54328434262|
|          103100.0|125919.93561440287|
|           92500.0| 160993.5841990849|
|          128100.0| 224169.1449215077|
|           99600.0| 186035.4703376263|
+------------------+------------------+
only showing top 20 rows



### 6.2 Модель 2 (только числовые переменные)

__Комметарий:__ Отделим числовые признаки и заново обучим модель

In [17]:
assembler_num = VectorAssembler(
    inputCols=['scaled_numerical_features'],  # ← ВАЖНО: используем МАСШТАБИРОВАННЫЕ признаки!
    outputCol="features_num"  
)

train_df = assembler_num.transform(train_df)
test_df = assembler_num.transform(test_df)

lr_2=LinearRegression(labelCol=target, featuresCol='features_num')
model_2=lr_2.fit(train_df)

pred_2=model_2.transform(test_df)
predicted_labels_2=pred_2.select('median_house_value', "prediction").withColumnRenamed("prediction", "pred_2")

predicted_labels_2.show(10)



26/01/30 14:13:13 WARN Instrumentation: [0ac5a26a] regParam is zero, which might cause numerical instability and overfitting.


+------------------+------------------+
|median_house_value|            pred_2|
+------------------+------------------+
|          103600.0|104470.30387285385|
|           50800.0|189334.43952907037|
|           58100.0|107973.21667532106|
|           68400.0| 90791.92265174419|
|           72200.0|130910.57959356006|
|           67000.0| 125852.1335151167|
|           81300.0|117885.71931501028|
|           70500.0|125710.89256039409|
|           60000.0| 105240.5449118526|
|          109400.0|122651.84458050603|
+------------------+------------------+
only showing top 10 rows



### 6.3 Модель 3 (Random Forest)

__Комментарий:__ Удалим объединенные признаки и объединим их заново 

In [18]:
train_df = train_df.drop("all_features")
test_df = test_df.drop("all_features")

train_df = assembler_all.transform(train_df)
test_df = assembler_all.transform(test_df)

In [19]:
rf = RandomForestRegressor(
    featuresCol="all_features",      # столбец с признаками
    labelCol="median_house_value",  # целевая переменная
    predictionCol="prediction",  # столбец для предсказаний 
    numTrees=100,                # количество деревьев
    maxDepth=10,                 # максимальная глубина деревьев
    seed=42                      # seed для воспроизводимости
)

rf_model = rf.fit(train_df)

pred_3 = rf_model.transform(test_df)
predicted_labels_3=pred_3.select('median_house_value', "prediction").withColumnRenamed("prediction", "pred_3")

predicted_labels_3.show(10)

26/01/30 14:13:29 WARN DAGScheduler: Broadcasting large task binary with size 1022.0 KiB
26/01/30 14:13:32 WARN DAGScheduler: Broadcasting large task binary with size 1954.0 KiB
26/01/30 14:13:35 WARN DAGScheduler: Broadcasting large task binary with size 3.7 MiB
26/01/30 14:13:37 WARN DAGScheduler: Broadcasting large task binary with size 1061.5 KiB
26/01/30 14:13:39 WARN DAGScheduler: Broadcasting large task binary with size 6.7 MiB
26/01/30 14:13:41 WARN DAGScheduler: Broadcasting large task binary with size 1776.7 KiB
26/01/30 14:13:44 WARN DAGScheduler: Broadcasting large task binary with size 11.7 MiB
26/01/30 14:13:48 WARN DAGScheduler: Broadcasting large task binary with size 2.7 MiB


+------------------+------------------+
|median_house_value|            pred_3|
+------------------+------------------+
|          103600.0|114499.61070698607|
|           50800.0|152053.03981000668|
|           58100.0| 89247.89338859332|
|           68400.0|109758.95488121333|
|           72200.0|100149.55279914316|
|           67000.0| 85976.10439208531|
|           81300.0| 97468.47945238918|
|           70500.0| 88686.03055455806|
|           60000.0|117654.74887241075|
|          109400.0| 138371.1366479284|
+------------------+------------------+
only showing top 10 rows



### 6.4 Анализ результатов

In [26]:
evaluator_1=RegressionEvaluator(labelCol="median_house_value", predictionCol="pred_1")

rmse_1=evaluator_1.evaluate(predicted_labels_1, {evaluator_1.metricName:"rmse"})
mae_1=evaluator_1.evaluate(predicted_labels_1, {evaluator_1.metricName:"mae"})
r2_1=evaluator_1.evaluate(predicted_labels_1, {evaluator_1.metricName:"r2"})

evaluator_2=RegressionEvaluator(labelCol="median_house_value", predictionCol="pred_2")

rmse_2=evaluator_2.evaluate(predicted_labels_2, {evaluator_2.metricName:"rmse"})
mae_2=evaluator_2.evaluate(predicted_labels_2, {evaluator_2.metricName:"mae"})
r2_2=evaluator_2.evaluate(predicted_labels_2, {evaluator_2.metricName:"r2"})

evaluator_3=RegressionEvaluator(labelCol="median_house_value", predictionCol="pred_3")

rmse_3=evaluator_3.evaluate(predicted_labels_3, {evaluator_3.metricName:"rmse"})
mae_3=evaluator_3.evaluate(predicted_labels_3, {evaluator_3.metricName:"mae"})
r2_3=evaluator_3.evaluate(predicted_labels_3, {evaluator_3.metricName:"r2"})

print("Для первой модели:")
print(f" ")
print(f"RMSE: {rmse_1}")
print(f"MAE: {mae_1}")
print(f"R2: {r2_1}")
print(f"____________________")
print(f" ")
print("Для второй модели:")
print(f" ")
print(f"RMSE: {rmse_2}")
print(f"MAE: {mae_2}")
print(f"R2: {r2_2}")
print(f"____________________")
print(f" ")
print("Для третьей модели:")
print(f" ")
print(f"RMSE: {rmse_3}")
print(f"MAE: {mae_3}")
print(f"R2: {r2_3}")



Для первой модели:
 
RMSE: 95572.92305986027
MAE: 53931.88786336528
R2: 0.325336621088518
____________________
 
Для второй модели:
 
RMSE: 71839.9154724157
MAE: 53059.21226278419
R2: 0.6188033946225383
____________________
 
Для третьей модели:
 
RMSE: 56902.04894986041
MAE: 39835.340208552
R2: 0.7608484450309494


__Сравнение__

In [29]:
dif_rmse=rmse_2-rmse_3
dif_mae=mae_2-mae_3
dif_r2=r2_3-r2_2

print(f"Разница RMSE: {dif_rmse}")
print(f"Разница MAE: {dif_mae}")
print(f"Разница R2: {dif_r2}")

Разница RMSE: 14937.86652255529
Разница MAE: 13223.87205423219
Разница R2: 0.1420450504084111


_______________________________________________________________________

<h1 style="font-size: 20px; color: #000000; font-weight: bold;">
    Вывод по разделу
</h1>
  
  
  
**Обучение моделей и предсказания**

`6.1 Модель 1: Линейная регрессия (все признаки)`
    
- **Архитектура**: полная линейная регрессия со всеми признаками
- **Признаки**: 8 числовых (масштабированных) + 5 категориальных (one-hot)
- **Предупреждение**: `regParam is zero` - отсутствует регуляризация, возможен overfitting
- **Первые предсказания**: видна значительная вариативность прогнозов

`6.2 Модель 2: Линейная регрессия (только числовые признаки)`
    
- **Архитектура**: линейная регрессия без категориальных признаков
- **Признаки**: только 8 числовых признаков
- **Аналогичное предупреждение**: отсутствие регуляризации
- **Наблюдение**: предсказания более консервативные по сравнению с моделью 1

`6.3 Модель 3: Random Forest (все признаки)`
    
- **Архитектура**: ансамблевая модель из 100 деревьев глубиной до 10
- **Особенности**: устойчива к переобучению, автоматически определяет важность признаков
- **Технические предупреждения**: связаны с broadcast больших данных в распределенной системе
- **Производительность**: требует больше вычислительных ресурсов

`6.4 Анализ результатов`

**Сравнение метрик по моделям:**

 ============================================================  
|                         Модель                          |  
============================================================  
| Линейная регрессия (все признаки)                       |
| RMSE: $95,572.92,  MAE: $53,931.89,  R²: 0.3253          |  
------------------------------------------------------------  
| Линейная регрессия (числовые)                           |
| RMSE: $71,839.92,  MAE: $53,059.21,  R²: 0.6188          |  
------------------------------------------------------------  
| Random Forest (все признаки)                            |
| RMSE: $56,902.05,  MAE: $39,835.34,  R²: 0.7608          |  
============================================================  

**Ключевые выводы:**

1. **Лучшая модель**: Random Forest показала наилучшие результаты по всем метрикам
    
   - **RMSE**: на 56,902$ ниже, чем у модели 1 и на 14,938$ ниже, чем у модели 2
   - **MAE**: ошибка снижена на 14,097$ относительно модели 1 и на 13,224$ относительно модели 2
   - **R²**: объясняет 76.08% дисперсии против 61.88% у модели 2

2. **Неожиданный результат**: 
    
   - Модель 1 (со всеми признаками) показала **худшие** результаты, чем модель 2 (только числовые признаки)
   

3. **Эффективность Random Forest**:
    
   - Улучшение R²: +0.1420 относительно модели 2
   - Снижение RMSE: на 26.3% относительно модели 1
   - Стабильность: лучшая устойчивость к выбросам (низкий MAE)

4. **Бизнес-интерпретация**:
    
   - **Наилучшая модель (Random Forest)**: ошибается в среднем на 39,835$, что составляет ~19% от средней стоимости дома
   


## 7. __Общий вывод по проекту__



<div style="border:solid blue 3px; padding: 20px">

`Подготовка данных:`
    
<strong>Итоги:</strong>
<ul>
<li><strong>Качество данных:</strong> Высокое, минимальное количество пропусков (207 из 20640 записей в total_bedrooms, ~1%)</li>
<li><strong>Обработка пропусков:</strong> Заполнены средним значением (537.87)</li>
<li><strong>Кодирование категориальных признаков:</strong> One-hot encoding для ocean_proximity с handleInvalid="keep"</li>
<li><strong>Масштабирование:</strong> StandardScaler применен ко всем числовым признакам</li>
<li><strong>Разделение данных:</strong> 80/20 на обучающую и тестовую выборки с фиксированным seed</li>
</ul>

`Обучение и сравнение моделей:`
    
<strong>Итоги:</strong>
<ul>
<li><strong>Модель 1 (Линейная регрессия - все признаки):</strong> RMSE: $95,572.92, MAE: $53,931.89, R²: 0.3253</li>
<li><strong>Модель 2 (Линейная регрессия - только числовые):</strong> RMSE: $71,839.92, MAE: $53,059.21, R²: 0.6188</li>
<li><strong>Модель 3 (Random Forest - все признаки):</strong> RMSE: $56,902.05, MAE: $39,835.34, R²: 0.7608 ★ НАИЛУЧШАЯ</li>
</ul>

`Ключевые выводы:`
<ul>
<li><strong>Random Forest превосходит линейные модели</strong> по всем метрикам:
  <ul>
  <li>Улучшение RMSE: на 40.5% относительно Модели 1</li>
  <li>Улучшение MAE: на 26.1% относительно Модели 1</li>
  <li>Объясняет 76.1% дисперсии против 61.9% у Модели 2</li>
  </ul>
</li>
<li><strong>Неожиданный результат:</strong> Линейная регрессия со всеми признаками показала худшие результаты, чем только с числовыми
  <ul>
  <li>Возможные причины: переобучение, мультиколлинеарность после one-hot кодирования, отсутствие регуляризации</li>
  </ul>
</li>
<li><strong>Бизнес-интерпретация:</strong> Наилучшая модель ошибается в среднем на $39,835, что составляет ~19% от средней стоимости дома</li>
</ul>

`Бизнес-ценность проекта:`
<ul>
<li><strong>Для риелторов:</strong> Автоматизация оценки объектов с точностью ~80%</li>
<li><strong>Для инвесторов:</strong> Выявление перспективных районов для вложений</li>
<li><strong>Для банков:</strong> Оценка залоговой стоимости при ипотечном кредитовании</li>
<li><strong>Для муниципалитетов:</strong> Анализ рынка жилья и планирование развития районов</li>
</ul>

`Заключение:`

Проект успешно решил задачу прогнозирования стоимости жилья в Калифорнии с использованием трех различных подходов. Наилучшие результаты показала модель Random Forest, которая объясняет 76.1% дисперсии целевой переменной со средней ошибкой $39,835. Модель готова для практического применения в реальных бизнес-сценариях оценки недвижимости.

<strong>Ключевая ценность:</strong> Создание автоматизированной системы оценки стоимости жилья с измеримой точностью, снижающей субъективность человеческой оценки и ускоряющей процесс принятия решений в сфере недвижимости.

</div>

__Уважаемый читатель, благодарю за уделенные время и внимание моему проекту!__